File: 2021_Samsung_SDS_algorithm_lecture/using_Cpp/1753.cpp
/*
input :
5 6
1
5 1 1
1 2 2
1 3 3
2 3 4
2 4 5
3 4 6

output :
0
2
3
7
INF
*/

#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

int V, E;
int S;
int a, b, w;
vector<pair<int, int>> graph[20002];
int dist[20002];
int INF = 4000000;


void dijkstra(int S){
    priority_queue<pair<int, int>> min_heap;
    dist[S] = 0;
    min_heap.push({0, S});

    while (!min_heap.empty()){
        pair<int, int> cur;
        cur = min_heap.top(); min_heap.pop();
        int cur_w = cur.first; int cur_node = cur.second;

        if (dist[cur_node] < - cur_w) continue;

        for (pair<int, int> nxt : graph[cur_node]){
            int nxt_w = nxt.first; int nxt_node = nxt.second;
            int new_cost = dist[cur_node] + nxt_w;
            if (new_cost > dist[nxt_node]) continue;

            dist[nxt_node] = new_cost;
            min_heap.push({-new_cost, nxt_node});
        }
    }
}


int main(){
    ios::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);
    freopen("input.txt", "r", stdin);

    cin >> V >> E;
    cin >> S;

    for (int i = 0; i < E; i++){
        cin >> a >> b >> w;
        graph[a].push_back({w, b});
    }
    for (int i = 1; i < V + 1; i++){
        dist[i] = INF;
    }

    dijkstra(S);
    for (int i = 1; i < V + 1; i++){
        if (dist[i] == INF)
            cout << "INF" << '\n';
        else
            cout << dist[i] << '\n';
    }
    return 0;
}

File: 2021_Samsung_SDS_algorithm_lecture/using_Python/1753.py
"""
input :
5 6
1
5 1 1
1 2 2
1 3 3
2 3 4
2 4 5
3 4 6

output :
0
2
3
7
INF
"""

import sys
import heapq
sys.stdin = open("input.txt", "r")
input = sys.stdin.readline


INF = 10**8


def dijkstra(S):
    min_heap = []
    dist[S] = 0
    heapq.heappush(min_heap, [0, S])

    while len(min_heap) != 0:
        cur_w, cur = heapq.heappop(min_heap)
        if dist[cur] < cur_w: continue

        for w, nxt in graph[cur]:
            new_cost = dist[cur] + w
            if new_cost > dist[nxt]: continue

            dist[nxt] = new_cost
            heapq.heappush(min_heap, [new_cost, nxt])


if __name__ == "__main__":
    V, E = map(int, input().rstrip().split())
    S = int(input())

    graph = [[] for _ in range(V + 1)]
    for _ in range(E):
        a, b, w = map(int, input().rstrip().split())
        graph[a].append([w, b])

    dist = [INF for _ in range(V + 1)]

    dijkstra(S)

    for i in range(1, V + 1):
        if dist[i] != INF:
            print(dist[i])
        else:
            print("INF")
